// @worktile/gantt/esm2020/table/gantt-table.component.mjs
import { Component, ContentChild, EventEmitter, Output } from '@angular/core';
import * as i0 from "@angular/core";
export class NgxGanttTableComponent {
    constructor() {
        this.columnChanges = new EventEmitter();
    }
}
NgxGanttTableComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.9", ngImport: i0, type: NgxGanttTableComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
NgxGanttTableComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.3.9", type: NgxGanttTableComponent, selector: "ngx-gantt-table", outputs: { columnChanges: "columnChanges" }, queries: [{ propertyName: "rowBeforeTemplate", first: true, predicate: ["rowBeforeSlot"], descendants: true, static: true }, { propertyName: "rowAfterTemplate", first: true, predicate: ["rowAfterSlot"], descendants: true, static: true }], ngImport: i0, template: '', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.9", ngImport: i0, type: NgxGanttTableComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'ngx-gantt-table',
                    template: ''
                }]
        }], propDecorators: { columnChanges: [{
                type: Output
            }], rowBeforeTemplate: [{
                type: ContentChild,
                args: ['rowBeforeSlot', { static: true }]
            }], rowAfterTemplate: [{
                type: ContentChild,
                args: ['rowAfterSlot', { static: true }]
            }] } });
