// @worktile/gantt/esm2020/gantt.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxGanttComponent } from './gantt.component';
import { NgxGanttTableComponent } from './table/gantt-table.component';
import { NgxGanttTableColumnComponent } from './table/gantt-column.component';
import { GanttCalendarComponent } from './components/calendar/calendar.component';
import { GanttTableComponent } from './components/table/gantt-table.component';
import { NgxGanttBarComponent } from './components/bar/bar.component';
import { GanttMainComponent } from './components/main/gantt-main.component';
import { GanttIconComponent } from './components/icon/icon.component';
import { GanttDragBackdropComponent } from './components/drag-backdrop/drag-backdrop.component';
import { GanttLinksComponent } from './components/links/links.component';
import { NgxGanttRootComponent } from './root.component';
import { NgxGanttRangeComponent } from './components/range/range.component';
import { IsGanttRangeItemPipe, IsGanttBarItemPipe, IsGanttCustomItemPipe } from './gantt.pipe';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { GANTT_GLOBAL_CONFIG, defaultConfig } from './gantt.config';
import { NgxGanttBaselineComponent } from './components/baseline/baseline.component';
import * as i0 from "@angular/core";
export class NgxGanttModule {
}
NgxGanttModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.9", ngImport: i0, type: NgxGanttModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxGanttModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.9", ngImport: i0, type: NgxGanttModule, declarations: [NgxGanttComponent,
        NgxGanttTableComponent,
        NgxGanttTableColumnComponent,
        GanttTableComponent,
        GanttMainComponent,
        GanttCalendarComponent,
        GanttLinksComponent,
        NgxGanttBarComponent,
        GanttIconComponent,
        GanttDragBackdropComponent,
        NgxGanttRangeComponent,
        NgxGanttRootComponent,
        NgxGanttBaselineComponent,
        IsGanttRangeItemPipe,
        IsGanttBarItemPipe,
        IsGanttCustomItemPipe], imports: [CommonModule, DragDropModule], exports: [NgxGanttComponent,
        NgxGanttTableComponent,
        NgxGanttTableColumnComponent,
        NgxGanttRootComponent,
        NgxGanttBarComponent,
        NgxGanttRangeComponent,
        NgxGanttBaselineComponent] });
NgxGanttModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.9", ngImport: i0, type: NgxGanttModule, providers: [
        {
            provide: GANTT_GLOBAL_CONFIG,
            useValue: defaultConfig
        }
    ], imports: [[CommonModule, DragDropModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.9", ngImport: i0, type: NgxGanttModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, DragDropModule],
                    exports: [
                        NgxGanttComponent,
                        NgxGanttTableComponent,
                        NgxGanttTableColumnComponent,
                        NgxGanttRootComponent,
                        NgxGanttBarComponent,
                        NgxGanttRangeComponent,
                        NgxGanttBaselineComponent
                    ],
                    declarations: [
                        NgxGanttComponent,
                        NgxGanttTableComponent,
                        NgxGanttTableColumnComponent,
                        GanttTableComponent,
                        GanttMainComponent,
                        GanttCalendarComponent,
                        GanttLinksComponent,
                        NgxGanttBarComponent,
                        GanttIconComponent,
                        GanttDragBackdropComponent,
                        NgxGanttRangeComponent,
                        NgxGanttRootComponent,
                        NgxGanttBaselineComponent,
                        IsGanttRangeItemPipe,
                        IsGanttBarItemPipe,
                        IsGanttCustomItemPipe
                    ],
                    providers: [
                        {
                            provide: GANTT_GLOBAL_CONFIG,
                            useValue: defaultConfig
                        }
                    ]
                }]
        }] });
